### PR TITLE
Feature/push callback

### PR DESCRIPTION
**Pull request by guperrot**

### CODE REVIEW

**Comment by guperrot** [comment]
The don't merge is just to avoid releasing that for now, please review nonetheless.

**Comment by dhei** [comment]
Looks good to me.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java [hunk 1]
```diff
@@ -224,4 +273,102 @@ public synchronized void setInstanceEnabled(boolean enabled) {
...
+                    deliverForegroundPushNotification(pushNotification);
+                }
+            });
+        }
+    }
+
+    /**
+     * Top level method needed for synchronized code coverage.
+     */
+    private synchronized void deliverForegroundPushNotification(PushNotification pushNotification) {
```
Comment (jaelim-ms) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: Just curious, any reason of taking out as a method? Looks it is used by one place.

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/PushListener.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+public interface PushListener {
+
+    /**
+     * Called whenever a push notification is either clicked from system notification center or
+     * when the push is received in foreground.
+     *
+     * @param activity         current activity when push is received or clicked.
+     * @param pushNotification the push notification details. If clicked from background, title and
+     *                         message will be empty.
+     *                         <p>
```
Comment (jaelim-ms) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/PushListener.java]: Is this necessary?

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java (hunk 1 above)
Comment (jaelim-ms) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: I missed synchronized.

Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java]: Response in the comment ;)

File: sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/PushListener.java (hunk 2 above)
Comment (guperrot) [inline sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/PushListener.java]: For better readability.

**Pull request merged**